opendkim/opendkim-spam.1.in
.TH opendkim-spam 1 "The Trusted Domain Project"
.SH NAME
.B opendkim-spam
\- DKIM filter spam correlation tool
.SH SYNOPSIS
.B opendkim-spam
[options]
.SH DESCRIPTION
.B opendkim-spam
accepts a regular format message (RFC5322) on standard input and uses it
to update a local SQL database being updated by
.I opendkim(8)
with an indiciation that a user believes the input message is spam or
otherwise abusive.  This feedback is important input toward developing
DKIM-based domain reputation systems.

The tool is intended to be used directly from within shell-based mail readers
such as
.I alpine(1)
or
.I mutt(1)
using a "pipe" command, which feeds the message being read to the specified
program.

This tool is experimental.  If the experiment proves useful, the
feedback could be used as an input stream to a larger-scale collaborative
feedback system that can be used to identify sources of signed mail
that have good reputations.
.SH OPTIONS
.TP
.I -b backend
Specifies the style of backend database in use.  The default is
"@SQL_BACKEND@".
.TP
.I -c file
Names a configuration file from which operating parameters will be read.
The configuration file contains entries of the form "key value", one per
line.  Empty lines or lines beginning with a hash ("#") character are
ignored.  Command line equivalents for the configuration file are available
as runtime overrides.
The default configuration file location is
.I @SYSCONFDIR@/opendkim-spam.conf.
See the CONFIGURATION FILE section for a list of known values.
.TP
.I -f
Run in the foreground rather than executing as a background process.
Intended mainly for debugging.
.TP
.I -d name
Indicates the name of the database to be accessed.  The default is
"opendkim".
.TP
.I -h host
Specifies the host where the database server is running.  The default
is "localhost".
.TP
.I -o file
Writes an update record to an OpenDKIM statistics file rather than directly
to a database.  There is no default.  Note that if this option is used,
all database-specific options are ignored.
.TP
.I -p password
Specifies the password to be used when authenticating to the database.
The default is "opendkim".
.TP
.I -P port
Specifies the port number where the database server is listening.  The
default depends on which backend is in use.
.TP
.I -r reporter
Provides the name of the site reporting the spam.  If not provided, an attempt
will be made to extract this information from the topmost Received header
field from the input message.
.TP
.I -s column
Names the database column whose value should be incremented as a result
of this user action.  The default is "spam".
.TP
.I -u user
Identifies the database user to be used when connecting to the database.
The default is "opendkim".
.TP
.I -v
Requests verbose output.  Can be specified multiple times for more and
more information.
.TP
.I -V
Print version number and exit.
.SH CONFIGURATION FILE
The configuration file used by
.B opendkim-spam(1)
is expected to be a text file.  Empty lines or lines starting with a
hash ("#") character are ignored.  All other lines should consist of a
parameter name followed by one or more whitespace characters, then followed
by its intended value.

Parameters generally match command line options (specified above), but when
present the command line options override the configuration file options.  The
list of configuration file parameters and their command line equivalents
are as follows; see above for descriptions:
.TP
.I Background
\-f (opposite meaning)
.TP
.I DatabaseBackend
\-b
.TP
.I DatabaseName
\-d
.TP
.I DatabaseHost
\-h
.TP
.I DatabasePassword
\-p
.TP
.I DatabasePort
\-P
.TP
.I DatabaseSpamColumn
\-s
.TP
.I DatabaseUser
\-u
.TP
.I ReporterID
\-r
.TP
.I SkipReceived
An integer that defines a number of Received fields that should be skipped
while searching for the one that was also seen by
.B opendkim(8)
so that this command refers to the same Received field.  Defaults to 0.
.TP
.I StatisticsFile
\-o
.SH NOTES
It is possible to compile this application without SQL support, in which
case only the
.I StatisticsFile
setting has any meaning (and, in fact, it is required in that case).
.SH VERSION
This man page covers the version of
.I opendkim-spam
that shipped with version @VERSION@ of
.I OpenDKIM.
.SH COPYRIGHT
Copyright (c) 2011, 2012, The Trusted Domain Project.  All rights reserved.
.SH SEE ALSO
.I alpine(1),
.I mutt(1),
.I opendkim(8)
.P
RFC6376 - DomainKeys Identified Mail


RELEASE_NOTES
			OPENDKIM RELEASE NOTES

This listing shows the versions of the OpenDKIM package, the date of
release, and a summary of the changes in that release.

2.11.0		2018/11/??
	Add support for ED25519 keys.  GNUTLS support provided by
		Dilyan Palauzov.
	Add "header.s" and "header.a" in generated Authentication-Results
		header fields.
	Add _FFR_CONDITIONAL, experimental support for conditional DKIM
		signatures.
	Make odkim.internal_ip() available to all Lua hooks.  Problem noted
		by Iosif Fettich.
	Make bind code DNSSEC-aware.  Patch from Jack Bates.
	Extend KeyTable to specify signing algorithm.
		Patch from Yasuhito Futatsuki.
	Fix dkimf_db_nextpunct() so it doesn't incorrectly identify an encoded
		hex digit as a value delimiter.
	Fix issue #8: The password file critical section isn't big enough.
		Patch from Dilyan Palauzov.
	Fix issue #9: Plug a few failure mode memory leaks.  Patch from
		Dilyan Palauzov.
	Fix issue #15: Don't automatically skip the body when one mode (sign
		or verify) doesn't need it, because the other one might.
		Patch from Dilyan Palauzov.
	Fix issue #28: RequiredHeaders should report and log the specific
		error and reject the message, as documented.  Patch from
		Dilyan Palauzov.
	Fix issue #222: When the "Minimum" option is set, don't treat
		empty body as partial. Patch from Yasuhito FUTATSUKI
		(PR #223).
	Fix bug #234: Make "NoHeaderB" do something.  Reported by Joseph
		Coffland.
	Fix bug #235: Quote "header.b" values in case they contain a slash
		(which must be quoted).
	Fix bug #237: Fix processing of "SoftStart".
	CONFIG: Add compatibility with openssl-1.1.0.  Patch from
		Scott Kitterman.
	CONTRIB: Simplfy some logic in contrib/repute.  Patch from
		Dilyan Palauzov.
	LIBOPENDKIM: Feature request #190: Reject signature object requests
		where the domain name or selector includes non-printable
		characters.  Suggested by Franck Martin.
	LIBOPENDKIM: Re-fix bug #226: Deal with header fields that are
		wrapped before there's any content.  Somehow the wrong fix
		was committed.  Originally reported by Alessandro Vesely;
		re-reported by David Stevenson.
	LIBOPENDKIM: Fix bug #233: In the weird case of duplicate signatures,
		constrain the size of the resulting "header.b" value.  Problem
		noted by Joseph Coffland.
	LIBOPENDKIM: Fix bug #270: Don't set an upper bound on the size of
		the canonicalization buffer, which interferes with processing
		oversized input lines.  Reported by Mars Peng.
	LIBOPENDKIM: Fix parsing bug in dkim_mail_parse_multi(), where quotes
		were not being properly handled.
	LIBOPENDKIM: Expose conversion table between internal code already
		provided as DKIM_ macros and their literal name in C string.
	LIBOPENDKIM: Extend dkim_test_key() to allow testing a ed25519 key,
		as dkim_test_key2(). For API compatibility, function interface
		of dkim_test_key() is not changed, but it only calls
		dkim_test_key2(). Patch from Yasuhito Futatsuki.
	TOOLS: Feature requrest #187: Add option to match subdomains when
		generating zone files.  Patch from Andreas Schulze.
	TOOLS: issue #183: On opendkim-testkey, add support for ed25519 keys.
		Patch from Yasuhito Futatsuki.

2.10.3		2015/05/12
	LIBOPENDKIM: Make strict header checking non-destructive.  The last
		change to this code resulted in failing signatures.  Reported
		by Pedro Morales and Wez Furlong.

2.10.2		2015/05/10
	Fix bug #221: Report a DKIM result of "policy" if MinimumKeyBits
		or UnprotectedKey cause the signature to result in a "pass"
		override.  Reported by Kurt Roeckx.
	Fix bug #227: Revert removal of SenderHeaders configuration setting.
		Document that it is now limited to signature selection.
	LIBOPENDKIM: Fix bug #226: Deal with header fields that are
		wrapped before there's any content.  Reported by
		Alessandro Vesely.
	CONTRIB: Update to contrib/systemd/opendkim.service.in from
		Steve Jenkins.

2.10.1		2015/02/03
	Fix bug #214: Handle arbitrarily large From: fields.  Reported by
		Tomohiko Sasaki.
	Fix bug #220: Make DB_SIGNINGTABLE symbol available in Lua scripts.
		Problem noted by Klaus Heinrich.
	LIBOPENDKIM: Fix bug #213: Remove "dkim_default_senderhdrs" from
		dkim.h.  Problem noted by Daniel J. Luke.
	LIBOPENDKIM: Fix bug #219: Unresolved CNAMEs are not failures,
		according to the DNS (see RFC6604), so report them as
		NXDOMAIN or similar.  Reported by Alessandro Vesely.

2.10.0		2014/12/27
	Feature request #182: Remove "AddAllSignatureResults".  All signature
		results will now be added via Authentication-Results header
		fields.  Requested by Tomki Camp.
	Feature request #180: Rename "LDAPSoftStart" to "SoftStart" and apply
		it to SQL connections as well.  Requested by Daniel Kauffman.
	Feature request #179: Add "IgnoreMalformedMail" option.
	Fix bug #183: Discontinue support for ADSP.  This removes the
		following configuration file items:
		AddAllSignatureResults	LocalADSP
		ADSPAction		NoDiscardableMailTo
		ADSPNoSuchDomain	On-PolicyError
		BogusPolicy		SendADSPReports
		DisableADSP		SenderHeaders
		LDAPSoftStart		UnprotectedPolicy
	Make "rrvs" and "smime" recognized Authentication-Results methods.
	LIBOPENDKIM: Feature request #157: Add dkim_mail_parse_multi().
		Suggested by Alessandro Vesely.
	LIBOPENDKIM: Feature request #185: Add dkim_set_dnssec().  Patch
		from Alec Peterson.
	LIBOPENDKIM: Fix bug #183: Discontinue support for ADSP.  This
		means all of the following:
		- the dkim_policy_t type has been removed
		- the DKIM_POLICY_* constants have been removed
		- the DKIM_PRESULT_* constants have been removed
		- passing DKIM_OPTS_SENDERHDRS to dkim_options() now
		  results in an error
		- the DKIM_PSTATE structure has been removed
		- all of the following functions have been removed:
		  dkim_policy(), dkim_policy_dnssec(),
		  dkim_policy_getqueries(), dkim_policy_getreportinfo(),
		  dkim_policy_state_free(), dkim_policy_state_new(),
		  dkim_policy_syntax(), dkim_getpolicystr(),
		  dkim_getpresult(), dkim_getpresultstr(),

opendkim/opendkim-spam.c
/*
**  Copyright (c) 2011-2013, 2015, The Trusted Domain Project.
**  	All rights reserved.
*/

#include "build-config.h"

/* system includes */
#include <sys/param.h>
#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>
#include <errno.h>
#include <ctype.h>
#include <stdlib.h>
#include <assert.h>
#ifdef HAVE_STDBOOL_H
# include <stdbool.h>
#endif /* HAVE_STDBOOL_H */

#ifdef USE_ODBX
/* opendbx includes */
# include <odbx.h>
#endif /* USE_ODBX */

/* opendkim includes */
#include "config.h"
#include "stats.h"

/* definitions, macros, etc. */
#define	BUFRSZ		1024
#define	CMDLINEOPTS	"b:c:d:fh:o:p:P:r:s:u:vV"
#define	DEFDBBACKEND	SQL_BACKEND
#undef	DEFCONFFILE
#define	DEFCONFFILE	CONFIG_BASE "/opendkim-spam.conf"
#define	DEFDBHOST	"localhost"
#define	DEFDBNAME	"opendkim"
#define	DEFDBPASS	"opendkim"
#define	DEFDBPORT	""
#define	DEFDBSPAMCOL	"spam"
#define	DEFDBUSER	"opendkim"
#define	MAXHEADER	16384
#define SPACES		"\r\n\t "

#ifndef FALSE
# define FALSE		0
#endif /* ! FALSE */
#ifndef TRUE
# define TRUE		1
#endif /* ! TRUE */

/* globals */
char *progname;

/* config definition */
struct configdef spam_config[] =
{
	{ "Background",			CONFIG_TYPE_BOOLEAN,	FALSE },
	{ "DatabaseBackend",		CONFIG_TYPE_STRING,	FALSE },
	{ "DatabaseHost",		CONFIG_TYPE_STRING,	FALSE },
	{ "DatabaseName",		CONFIG_TYPE_STRING,	FALSE },
	{ "DatabasePassword",		CONFIG_TYPE_STRING,	FALSE },
	{ "DatabaseSpamColumn",		CONFIG_TYPE_STRING,	FALSE },
	{ "DatabaseUser",		CONFIG_TYPE_STRING,	FALSE },
	{ "ReporterID",			CONFIG_TYPE_STRING,	FALSE },
	{ "SkipReceived",		CONFIG_TYPE_INTEGER,	FALSE },
	{ "StatisticsFile",		CONFIG_TYPE_STRING,	FALSE },
	{ NULL,				(u_int) -1,		FALSE }
};

/*
**  USAGE -- usage message
**
**  Parameters:
**  	None.
**
**  Return value:
**  	EX_USAGE
*/

int
usage(void)
{
	fprintf(stderr, "%s: usage: %s [options]\n"
	        "\t-b backend  \tdatabase backend [%s]\n"
	        "\t-c file     \tconfiguration file [%s]\n"
	        "\t-d dbname   \tdatabase name [%s]\n"
	        "\t-f          \trun in the foreground\n"
	        "\t-h dbhost   \tdatabase hostname [%s]\n"
	        "\t-o file     \tstatistics file\n"
	        "\t-p dbpass   \tdatabase password [%s]\n"
	        "\t-P dbport   \tdatabase port [%s]\n"
	        "\t-s dbspamcol\tdatabase spam column name [%s]\n"
	        "\t-u dbuser   \tdatabase user [%s]\n"
		"\t-r reporter \treporter id (reporting host name)\n"
	        "\t-v          \tbe more verbose\n"
	        "\t-V          \tprint version number and exit\n",
	        progname, progname,
	        DEFDBBACKEND,
	        DEFCONFFILE,
	        DEFDBNAME,
	        DEFDBHOST,
	        DEFDBPASS,
	        DEFDBPORT,
	        DEFDBSPAMCOL,
	        DEFDBUSER);

	return EX_USAGE;
}

/*
**  MAIN -- program mainline
**
**  Parameters:
**  	argc, argv -- the usual
**
**  Return value:
**  	Exit status.
*/

int
main(int argc, char **argv)
{
	_Bool dofork = TRUE;
	int c;
	int verbose = 0;
	int dberr;
	int repid;
	int msgid;
	u_int skipcount = 0;
	char *p;
	char *prev;
	char *dbbackend = NULL;
	char *dbuser = NULL;
	char *dbpass = NULL;
	char *dbname = NULL;
	char *dbhost = NULL;
	char *dbspamcol = NULL;
	char *dbport = NULL;
	char *conffile = DEFCONFFILE;
	char *job = NULL;
	char *reporter = NULL;
	char *statsfile = NULL;
	FILE *sf;
	struct config *conf = NULL;
#ifdef USE_ODBX
	odbx_t *db = NULL;
	odbx_result_t *result;
#endif /* USE_ODBX */
	char buf[BUFRSZ + 1];
	char rcvd[MAXHEADER + 1];

	progname = (p = strrchr(argv[0], '/')) == NULL ? argv[0] : p + 1;

	while ((c = getopt(argc, argv, CMDLINEOPTS)) != -1)
	{
		switch (c)
		{
		  case 'b':
			dbbackend = optarg;
			break;

		  case 'c':
			conffile = optarg;
			break;

		  case 'd':
			dbname = optarg;
			break;

		  case 'f':
			dofork = FALSE;
			break;

		  case 'h':
			dbhost = optarg;
			break;

		  case 'P':
			dbport = optarg;
			break;

		  case 'o':
			statsfile = optarg;
			break;

		  case 'p':
			dbpass = optarg;
			break;

		  case 's':
			dbspamcol = optarg;
			break;

		  case 'u':
			dbuser = optarg;
			break;

		  case 'r':
			reporter = optarg;
			break;

		  case 'v':
			verbose++;
			break;

		  case 'V':
			fprintf(stdout, "%s v%s\n", progname, VERSION);
			return EX_OK;

		  default:
			return usage();
		}
	}

	/* load from config file, if specified */
	if (conffile != NULL)
	{
		_Bool tmpf;
		unsigned int line;
		char path[MAXPATHLEN + 1];

		if (verbose >= 2)
		{
			fprintf(stdout, "%s: loading configuration from %s\n",
			        progname, conffile);
		}

		memset(path, '\0', sizeof path);
		conf = config_load(conffile, spam_config, &line,
		                   path, sizeof path, NULL);

		if (conf == NULL)
		{
			fprintf(stderr,
			        "%s: %s: configuration error at line %u: %s\n",
			        progname, path, line, config_error());
			return EX_CONFIG;

		}

		/* extract values */
		(void) config_get(conf, "SkipReceived",
		                  &skipcount, sizeof skipcount);

		if (dbbackend == NULL)
		{
			(void) config_get(conf, "DatabaseBackend",
			                  &dbbackend, sizeof dbbackend);
		}

		if (dbhost == NULL)
		{
			(void) config_get(conf, "DatabaseHost",
			                  &dbhost, sizeof dbhost);
		}

                if (dbname == NULL)
		{
			(void) config_get(conf, "DatabaseName",
			                  &dbname, sizeof dbname);
		}

		if (dbpass == NULL)
		{
			(void) config_get(conf, "DatabasePassword",
			                  &dbpass, sizeof dbpass);
		}

		if (dbport == NULL)
		{
			(void) config_get(conf, "DatabasePort",
			                  &dbport, sizeof dbport);
		}

		if (dbspamcol == NULL)
		{
			(void) config_get(conf, "DatabaseSpamColumn",
			                  &dbspamcol, sizeof dbspamcol);
		}

                if (dbuser == NULL)
		{
			(void) config_get(conf, "DatabaseUser",
			                  &dbuser, sizeof dbuser);
		}

		if (reporter == NULL)
		{
			(void) config_get(conf, "ReporterID",
			                  &reporter, sizeof reporter);
		}

		if (statsfile == NULL)
		{
			(void) config_get(conf, "StatisticsFile",
			                  &statsfile, sizeof statsfile);
		}

		if (config_get(conf, "Background", &tmpf, sizeof tmpf) == 1)
		{
			if (dofork && !tmpf)
				tmpf = FALSE;
		}
	}

	if (dbbackend == NULL)
		dbbackend = DEFDBBACKEND;
	if (dbhost == NULL)
		dbhost = DEFDBHOST;
	if (dbname == NULL)
		dbname = DEFDBNAME;
	if (dbpass == NULL)
		dbpass = DEFDBPASS;
	if (dbspamcol == NULL)
		dbspamcol = DEFDBSPAMCOL;
	if (dbuser == NULL)
		dbuser = DEFDBUSER;

	/* connect to the DB */
	if (statsfile != NULL)
	{
		sf = fopen(statsfile, "a");
		if (sf == NULL)
		{
			fprintf(stderr, "%s: %s: fopen(): %s\n", progname,
			        statsfile, strerror(errno));
			return EX_OSERR;
		}

		/* write version if file is new */
		if (ftell(sf) == 0)
			fprintf(sf, "V%d\n", DKIMS_VERSION);
	}
	else
	{
#ifndef USE_ODBX
		fprintf(stderr, "%s: SQL not supported in this installation\n",
		        progname);
		return EX_SOFTWARE;

#else /* ! USE_ODBX */

		dberr = odbx_init(&db, dbbackend, dbhost, dbport);
		if (dberr < 0)
		{
			fprintf(stderr, "%s: odbx_init(): %s\n", progname,
			        odbx_error(NULL, dberr));
			return EX_SOFTWARE;
		}

		if (verbose >= 1)
		{
			fprintf(stdout, "%s: connected to database on %s\n",
			        progname, dbhost);
		}

		dberr = odbx_bind(db, dbname, dbuser, dbpass,
		                  ODBX_BIND_SIMPLE);
		if (dberr < 0)
		{
			fprintf(stderr, "%s: odbx_bind(): %s\n", progname,
			        odbx_error(db, dberr));
			(void) odbx_finish(db);
			return EX_SOFTWARE;
		}

		if (verbose >= 2)
		{
			fprintf(stdout, "%s: database binding successful\n",
			        progname);
		}
#endif /* ! USE_ODBX */
	}

	/* read first Received:, extract reporter and job ID */
	memset(buf, '\0', sizeof buf);
	memset(rcvd, '\0', sizeof rcvd);
	while (fgets(buf, sizeof buf - 1, stdin) != NULL)
	{
		for (p = buf; *p != '\0'; p++)
		{
			if (*p == '\r' || *p == '\n')
			{
				*p = '\0';
				break;
			}
		}

		if (strncasecmp(buf, "Received:", 9) == 0)
		{
			if (skipcount > 0)
			{
				skipcount--;
				continue;
			}
			else
			{
				strlcpy(rcvd, buf, sizeof rcvd);
			}
		}
		else if (isascii(buf[0]) && isspace(buf[0]) &&
		         rcvd[0] != '\0')
		{
			strlcat(rcvd, buf, sizeof rcvd);
		}
		else if (isascii(buf[0]) && !isspace(buf[0]) &&
		         rcvd[0] != '\0')
		{
			break;
		}
	}

	if (rcvd[0] == '\0')
	{
		fprintf(stderr, "%s: Received header field not found\n",
		        progname);

#ifdef USE_ODBX
		if (db != NULL)
		{
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
		}
		else
		{
#endif /* USE_ODBX */
			fclose(sf);
#ifdef USE_ODBX
		}
#endif /* USE_ODBX */
		return EX_DATAERR;
	}

	/*
	**  This is overly simplistic as it doesn't account for CFWS or
	**  escapes, but for now it's a good starting point.
	*/

	prev = NULL;
	for (p = strtok(&rcvd[9], SPACES);
	     p != NULL;
	     p = strtok(NULL, SPACES))
	{
		if (prev != NULL)
		{
			if (strcasecmp(prev, "id") == 0)
				job = p;
			else if (reporter == NULL &&
			         strcasecmp(prev, "by") == 0)
				reporter = p;
		}

		prev = p;
	}

	if (job == NULL)
	{
		fprintf(stderr,
		        "%s: could not locate job ID in Received header field\n",
		        progname);

#ifdef USE_ODBX
		if (db != NULL)
		{
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
		}
		else
		{
#endif /* USE_ODBX */
			fclose(sf);
#ifdef USE_ODBX
		}
#endif /* USE_ODBX */
		return EX_DATAERR;
	}
	else if (reporter == NULL)
	{
		fprintf(stderr,
		        "%s: could not locate receiving host in Received header field\n",
		        progname);

#ifdef USE_ODBX
		if (db != NULL)
		{
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
		}
		else
		{
#endif /* USE_ODBX */
			fclose(sf);
#ifdef USE_ODBX
		}
#endif /* USE_ODBX */
		return EX_DATAERR;
	}

	for (p = job; *p != '\0'; p++)
	{
		if (*p == ';')
		{
			*p = '\0';
			break;
		}
	}

	if (sf != NULL)
	{
		fprintf(sf, "U%s\t%s\t0\t1\n", job, reporter);
		fclose(sf);
		return 0;
	}

#ifdef USE_ODBX
	if (verbose >= 1)
	{
		fprintf(stdout, "%s: requesting reporter id for '%s'\n",
		        progname, reporter);
	}

	if (dofork)
	{
		pid_t pid;

		pid = fork();
		switch (pid)
		{
		  case -1:
			fprintf(stderr, "%s: fork(): %s\n",
			        progname, strerror(errno));
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_OSERR;

		  case 0:
			(void) setsid();
			/* XXX -- should probably dup2() /dev/null here */
			break;

		  default:
			return 0;
		}
	}

	/* retrieve reporter ID */
	snprintf(buf, sizeof buf, "SELECT id FROM reporters WHERE name = '%s'",
	         reporter);
	if (verbose >= 3)
		fprintf(stdout, ">>> %s\n", buf);

	dberr = odbx_query(db, buf, 0);
	if (dberr != ODBX_ERR_SUCCESS)
	{
		fprintf(stderr, "%s: odbx_query(): %s\n", progname,
		        odbx_error(db, dberr));
		(void) odbx_unbind(db);
		(void) odbx_finish(db);
		return EX_SOFTWARE;
	}

	repid = -1;

	while ((dberr = odbx_result(db, &result, NULL, 0)) != ODBX_RES_DONE)
	{
		if (dberr < 0)
		{
			fprintf(stderr, "%s: odbx_result(): %s\n",
			        progname, odbx_error(db, dberr));
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;
		}

		switch (dberr)
		{
		  case ODBX_RES_TIMEOUT:
			fprintf(stderr, "%s: odbx_result(): timeout\n",
			        progname);
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;

		  case ODBX_RES_NOROWS:
			fprintf(stderr,
			        "%s: odbx_result(): unexpected return value\n",
			        progname);
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;

		  case ODBX_RES_ROWS:
			while ((dberr = odbx_row_fetch(result)) != ODBX_ROW_DONE)
			{
				if (dberr < 0)
				{
					fprintf(stderr,
					        "%s: odbx_row_fetch(): %s\n",
						progname,
					        odbx_error(db, dberr));
					(void) odbx_unbind(db);
					(void) odbx_finish(db);
					return EX_SOFTWARE;
				}

				if (repid != -1)
				{
					fprintf(stderr,
					        "%s: duplicate entries for repoter '%s'\n",
						progname,
					        reporter);
					(void) odbx_unbind(db);
					(void) odbx_finish(db);
					return EX_SOFTWARE;
				}

				repid = atoi(odbx_field_value(result, 0));
				if (repid <= 0)
				{
					fprintf(stderr,
					        "%s: unexpected reporter id '%s'\n",
						progname,
					        odbx_field_value(result, 0));
					(void) odbx_unbind(db);
					(void) odbx_finish(db);
					return EX_SOFTWARE;
				}
			}

			break;

		  default:
			assert(0);
		}
	}

	(void) odbx_result_finish(result);

	if (repid <= 0)
	{
		fprintf(stderr,
		        "%s: could not determine reporter id for '%s'\n",
		        progname, reporter);

		(void) odbx_unbind(db);
		(void) odbx_finish(db);
		return EX_DATAERR;
	}

	if (verbose >= 2)
		fprintf(stdout, "%s: reporter id = %d\n", progname, repid);

	if (verbose >= 1)
	{
		fprintf(stdout, "%s: requesting message id for '%s'\n",
		        progname, job);
	}

	/* get message ID */
	snprintf(buf, sizeof buf,
	         "SELECT MAX(id) FROM messages WHERE jobid = '%s' AND reporter = %d",
	         job, repid);
	if (verbose >= 3)
		fprintf(stdout, ">>> %s\n", buf);
	dberr = odbx_query(db, buf, 0);
	if (dberr != ODBX_ERR_SUCCESS)
	{
		fprintf(stderr, "%s: odbx_query(): %s\n", progname,
		        odbx_error(db, dberr));
		(void) odbx_unbind(db);
		(void) odbx_finish(db);
		return EX_SOFTWARE;
	}

	msgid = -1;
	while ((dberr = odbx_result(db, &result, NULL, 0)) != ODBX_RES_DONE)
	{
		if (dberr < 0)
		{
			fprintf(stderr, "%s: odbx_result(): %s\n",
			        progname, odbx_error(db, dberr));
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;
		}

		switch (dberr)
		{
		  case ODBX_RES_TIMEOUT:
			fprintf(stderr, "%s: odbx_result(): timeout\n",
			        progname);
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;

		  case ODBX_RES_NOROWS:
			fprintf(stderr,
			        "%s: odbx_result(): unexpected return value\n",
			        progname);
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;

		  case ODBX_RES_ROWS:
			while ((dberr = odbx_row_fetch(result)) != ODBX_ROW_DONE)
			{
				if (dberr < 0)
				{
					fprintf(stderr,
					        "%s: odbx_row_fetch(): %s\n",
						progname,
					        odbx_error(db, dberr));
					(void) odbx_unbind(db);
					(void) odbx_finish(db);
					return EX_SOFTWARE;
				}

				if (msgid != -1)
				{
					fprintf(stderr,
					        "%s: duplicate entries for job '%s' from reporter '%s'\n",
						progname,
					        job, reporter);
					(void) odbx_unbind(db);
					(void) odbx_finish(db);
					return EX_SOFTWARE;
				}


				msgid = atoi(odbx_field_value(result, 0));
				if (msgid <= 0)
				{
					fprintf(stderr,
					        "%s: unexpected message id '%s'\n",
						progname,
					        odbx_field_value(result, 0));
					(void) odbx_unbind(db);
					(void) odbx_finish(db);
					return EX_SOFTWARE;
				}
			}

			break;

		  default:
			assert(0);
		}
	}

	(void) odbx_result_finish(result);

	if (msgid <= 0)
	{
		fprintf(stderr,
		        "%s: could not determine message id for '%s'\n",
		        progname, job);

		(void) odbx_unbind(db);
		(void) odbx_finish(db);
		return EX_DATAERR;
	}

	if (verbose >= 2)
		fprintf(stdout, "%s: message id = %d\n", progname, msgid);

	/* issue update */
	if (verbose >= 1)
		fprintf(stdout, "%s: updating record\n", progname);

	snprintf(buf, sizeof buf,
	         "UPDATE messages SET %s = %s + 1 WHERE id = %d AND %s >= 0",
	         dbspamcol, dbspamcol, msgid, dbspamcol);
	if (verbose >= 3)
		fprintf(stdout, ">>> %s\n", buf);
	dberr = odbx_query(db, buf, 0);
	if (dberr != ODBX_ERR_SUCCESS)
	{
		fprintf(stderr, "%s: odbx_query(): %s\n", progname,
		        odbx_error(db, dberr));
		(void) odbx_unbind(db);
		(void) odbx_finish(db);
		return EX_SOFTWARE;
	}

	for (;;)
	{
		dberr = odbx_result(db, &result, NULL, 0);
		if (dberr == ODBX_RES_DONE || dberr == ODBX_RES_NOROWS)
		{
			break;
		}
		else if (dberr < 0)
		{
			fprintf(stderr, "%s: odbx_result(): %s\n",
			        progname, odbx_error(db, dberr));
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;
		}

		switch (dberr)
		{
		  case ODBX_RES_TIMEOUT:
			fprintf(stderr, "%s: odbx_result(): timeout\n",
			        progname);
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;

		  case ODBX_RES_ROWS:
			fprintf(stderr,
			        "%s: odbx_result(): unexpected return value\n",
			        progname);
			(void) odbx_unbind(db);
			(void) odbx_finish(db);
			return EX_SOFTWARE;

		  default:
			assert(0);
		}
	}

	(void) odbx_result_finish(result);

	if (verbose >= 1)
		fprintf(stdout, "%s: record updated\n", progname);

	/* close down */
	(void) odbx_unbind(db);
	(void) odbx_finish(db);
#endif /* USE_ODBX */

	return 0;
}
